controllers/serviceaccount_controller.go

```diff
@@ -249,74 +255,81 @@ func (r ServiceAccountReconciler) ReconcileNormal(guestClusterCtx *vmwarecontext
+		if err := r.ensureServiceAccount(ctx, pSvcAccount); err != nil {
 			return errors.Wrapf(err, "unable to create provider serviceaccount %s", pSvcAccount.Name)
 		}
 		// 2. Update configmap with serviceaccount
-		if err := r.ensureServiceAccountConfigMap(guestClusterCtx.ClusterContext, pSvcAccount); err != nil {
+		if err := r.ensureServiceAccountConfigMap(ctx, pSvcAccount); err != nil {
 			return errors.Wrapf(err, "unable to sync configmap for provider serviceaccount %s", pSvcAccount.Name)
 		}
 		// 3. Create secret of Service account token type for the service account
-		if err := r.ensureServiceAccountSecret(guestClusterCtx.ClusterContext, pSvcAccount); err != nil {
+		if err := r.ensureServiceAccountSecret(ctx, pSvcAccount); err != nil {
 			return errors.Wrapf(err, "unable to create provider serviceaccount secret %s", getServiceAccountSecretName(pSvcAccount))
 		}
 
 		// 4. Create the associated role for the service account
-		if err := r.ensureRole(guestClusterCtx.ClusterContext, pSvcAccount); err != nil {
+		if err := r.ensureRole(ctx, pSvcAccount); err != nil {
 			return errors.Wrapf(err, "unable to create role for provider serviceaccount %s", pSvcAccount.Name)
 		}
 
 		// 5. Create the associated roleBinding for the service account
-		if err := r.ensureRoleBinding(guestClusterCtx.ClusterContext, pSvcAccount); err != nil {
+		if err := r.ensureRoleBinding(ctx, pSvcAccount); err != nil {
 			return errors.Wrapf(err, "unable to create rolebinding for provider serviceaccount %s", pSvcAccount.Name)
 		}
 
 		// 6. Sync the service account with the target
-		if err := r.syncServiceAccountSecret(guestClusterCtx, pSvcAccount); err != nil {
+		if err := r.syncServiceAccountSecret(ctx, guestClusterCtx, pSvcAccount); err != nil {
 			return errors.Wrapf(err, "unable to sync secret for provider serviceaccount %s", pSvcAccount.Name)
 		}
 	}
 	return nil
 }
 
-func (r ServiceAccountReconciler) ensureServiceAccount(clusterCtx *vmwarecontext.ClusterContext, pSvcAccount vmwarev1.ProviderServiceAccount) error {
+func (r *ServiceAccountReconciler) ensureServiceAccount(ctx context.Context, pSvcAccount vmwarev1.ProviderServiceAccount) error {
+	log := ctrl.LoggerFrom(ctx)
+
 	svcAccount := corev1.ServiceAccount{
 		ObjectMeta: metav1.ObjectMeta{
 			Name:      getServiceAccountName(pSvcAccount),
 			Namespace: pSvcAccount.Namespace,
 		},
 	}
-	logger := clusterCtx.Logger.WithValues("providerserviceaccount", pSvcAccount.Name, "serviceaccount", svcAccount.Name)
-	err := util.SetControllerReferenceWithOverride(&pSvcAccount, &svcAccount, clusterCtx.Scheme)
+
+	err := util.SetControllerReferenceWithOverride(&pSvcAccount, &svcAccount, r.Client.Scheme())
 	if err != nil {
 		return err
 	}
-	logger.V(4).Info("Creating service account")
-	err = clusterCtx.Client.Create(clusterCtx, &svcAccount)
+	log.V(4).Info("Creating service account")
+	err = r.Client.Create(ctx, &svcAccount)
```

Comment:
This now uses the client to the management cluster instead of the one received from the cluster Tracker!

Reply:
Similar for other functions.

Reply:
The clusterCtx.Client was also a mgmt cluster client

Reply:
Please double check though. This would be a bad mistake (in fact I double checked exactly this because I made that mistake elsewhere during implementation)

Reply:
Jep you are right.
`ctx.GuestClient` is the only client to the guest cluster.